### PR TITLE
[CDAP-16612] Make sure correct classpath is used when using ssh for s…

### PR DESCRIPTION
**Pull request by CuriousVini**
…treaming pipeline launch

Also made launching jobs using dataproc job apis configurable through provisioner context.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java [hunk 1]
```diff
@@ -194,6 +194,10 @@ public final ProgramController run(final Program program, ProgramOptions oldOpti
                                                          cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile());
     try {
       final ProgramLaunchConfig launchConfig = new ProgramLaunchConfig();
+      if (clusterMode == ClusterMode.ON_PREMISE) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java]: Should it be ISOLATED?

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/DistributedSparkProgramRunner.java [hunk 2]
```diff
@@ -137,15 +140,24 @@ protected void setupLaunchConfig(ProgramLaunchConfig launchConfig, Program progr
...
+    Arguments userArguments = options.getUserArguments();
+    Map<String, String> clientArgs = RuntimeArguments.extractScope("task", "client", userArguments.asMap());
     // Add runnable. Only one instance for the spark client
     launchConfig.addRunnable(spec.getName(), new SparkTwillRunnable(spec.getName()), 1,
                              clientArgs, spec.getClientResources(), 0);
 
     Map<String, String> extraEnv = new HashMap<>();
     extraEnv.put(Constants.SPARK_COMPAT_ENV, sparkCompat.getCompat());
 
+    if (userArguments.hasOption(SystemArguments.RUNTIME_MANAGER_TYPE) &&
```
Comment (chtyim) [inline cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/DistributedSparkProgramRunner.java]: Why only for spark runner?

Comment (chtyim) [reply cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/DistributedSparkProgramRunner.java]: Also, it shouldn't matter of the launch type.

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/DistributedSparkProgramRunner.java [hunk 3]
```diff
@@ -137,15 +140,24 @@ protected void setupLaunchConfig(ProgramLaunchConfig launchConfig, Program progr
...
+    Arguments userArguments = options.getUserArguments();
+    Map<String, String> clientArgs = RuntimeArguments.extractScope("task", "client", userArguments.asMap());
     // Add runnable. Only one instance for the spark client
     launchConfig.addRunnable(spec.getName(), new SparkTwillRunnable(spec.getName()), 1,
                              clientArgs, spec.getClientResources(), 0);
 
     Map<String, String> extraEnv = new HashMap<>();
     extraEnv.put(Constants.SPARK_COMPAT_ENV, sparkCompat.getCompat());
 
+    if (!userArguments.hasOption(SystemArguments.RUNTIME_MANAGER_TYPE) ||
```
Comment (chtyim) [inline cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/DistributedSparkProgramRunner.java]: What is this for?

Comment (CuriousVini) [reply cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/DistributedSparkProgramRunner.java]: Had offline discussion about this condition. Made changes as per the discussion. Testing manually now.

Comment (CuriousVini) [reply cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/DistributedSparkProgramRunner.java]: Had offline discussion about this condition. Made changes as per the discussion. Testing manually now.

**Comment by CuriousVini** [comment]
Manually tested following scenarios for spark streaming pipeline:
1. Use dataproc apis to launch AND monitor job
2. Use ssh to launch AND monitor job
3. Use dataproc apis to launch job AND use ssh to monitor job.

**Comment by CuriousVini** [comment]
Manually tested spark pipeline for following scenarios:
1. Use dataproc apis to launch AND monitor job
2. Use ssh to launch AND monitor job
3. Use dataproc apis to launch job AND use ssh to monitor job.

**Review by chtyim — APPROVED** [review]

**Comment by CuriousVini** [comment]
build is green. Merging.

**Pull request merged**